src/shaders/AtmosphereProgram.js

```diff
@@ -0,0 +1,322 @@
+            this.scratchArray9 = new Float32Array(9);
+        };
+
+        /**
+         * A string that uniquely identifies this program.
+         * @type {string}
+         * @readonly
+         */
+        AtmosphereProgram.key = "WorldWindGpuAtmosphereProgram";
+
+        // Inherit from GpuProgram.
+        AtmosphereProgram.prototype = Object.create(GpuProgram.prototype);
+
+        /**
+         * Returns the atmosphere's altitude.
+         * @returns {Number} The atmosphere's altitude in meters.
+         */
+        AtmosphereProgram.prototype.getAltitude = function () {
+            return this.altitude;
+        };
+
+        /**
+         * Returns the scale depth.
+         * @returns {Number} The Rayleigh scale depth.
+         */
+        AtmosphereProgram.prototype.getScaleDepth = function () {
+            return this.rayleighScaleDepth;
+        };
+
+        /**
+         * Loads the specified number as the value of this program's 'fragMode' uniform variable.
+         *
+         * @param {WebGLRenderingContext} gl The current WebGL context.
+         * @param {Number} fragMode The frag mode value.
+         * @throws {ArgumentError} If the specified number is null or undefined.
+         */
+        AtmosphereProgram.prototype.loadFragMode = function (gl, fragMode) {
+            if (!fragMode) {
+                throw new ArgumentError(
+                    Logger.logMessage(Logger.LEVEL_SEVERE, "AtmosphereProgram", "loadFragMode", "missingFragMode"));
+            }
+
+            gl.uniform1i(this.fragModeLocation, fragMode);
+        };
+
+        /**
+         * Loads the specified matrix as the value of this program's 'mvpMatrix' uniform variable.
+         *
+         * @param {WebGLRenderingContext} gl The current WebGL context.
+         * @param {Matrix} matrix The matrix to load.
+         * @throws {ArgumentError} If the specified matrix is null or undefined.
+         */
+        AtmosphereProgram.prototype.loadModelviewProjection = function (gl, matrix) {
+            if (!matrix) {
+                throw new ArgumentError(
+                    Logger.logMessage(Logger.LEVEL_SEVERE, "AtmosphereProgram", "loadModelviewProjection",
+                        "missingMatrix"));
+            }
+
+            this.loadUniformMatrix(gl, matrix, this.mvpMatrixLocation);
+        };
+
+        /**
+         * Loads the specified vector as the value of this program's 'vertexOrigin' uniform variable.
+         *
+         * @param {WebGLRenderingContext} gl The current WebGL context.
+         * @param {Vec3} vector The vector to load.
+         * @throws {ArgumentError} If the specified vector is null or undefined.
+         */
+        AtmosphereProgram.prototype.loadVertexOrigin = function (gl, vector) {
+            if (!vector) {
+                throw new ArgumentError(
+                    Logger.logMessage(Logger.LEVEL_SEVERE, "AtmosphereProgram", "loadVertexOrigin", "missingVector"));
+            }
+
```

Comment:
It would be better to use uniform3f here instead of uniform3fv, thereby removing an unnecessary array allocation. Please apply that pattern to the remaining uniform code in AtmosphereProgram.